server/src/main/java/com/cloud/network/element/ConfigDriveNetworkElement.java

```diff
@@ -639,7 +639,10 @@ private boolean deleteConfigDriveIso(final VirtualMachine vm) throws ResourceUna
 
         if (location == Location.SECONDARY) {
             dataStore = _dataStoreMgr.getImageStoreWithFreeCapacity(vm.getDataCenterId());
-            agentId = findAgentIdForImageStore(dataStore);
+            if (!VirtualMachineManager.VmConfigDriveOnPrimaryPool.valueIn(vm.getDataCenterId()) &&
+                    !VirtualMachineManager.VmConfigDriveForceHostCacheUse.valueIn(vm.getDataCenterId()) && dataStore != null) {
+                agentId = findAgentIdForImageStore(dataStore);
+            }
```

Comment:
@Pearl1594 Any reason, why config drive deletion should depend on these settings _vm.configdrive.primarypool.enabled / vm.configdrive.use.host.cache.on.unsupported.pool_  (which are dynamic) ?

Reply:
@sureshanaparti if case the config drive iso for the vm is on primary store or image cache, then there is no need to search for image store agent id ..

Reply:
But, these settings doesn't indicate the VM config drive on primary store or host cache. What if these are changed after VM creation?

Reply:
> But, these settings doesn't indicate the VM config drive on primary store or host cache. What if these are changed after VM creation? Only when deploying a VM, this settings help to choose the config drive location.

Reply:
agree with @sureshanaparti 
it is better to check what do exist on primary, secondary or host.

Reply:
got it @sureshanaparti @weizhouapache - I'll handle it Thanks.